### DISCUSSION TITLE
Help with Configuring Custom GPIO Pins

### DISCUSSION THREAD

**Discussion by manuaatitya** [zephyrproject-rtos/zephyr#35932] [Q&A]
Hi, 
    I'm trying to write an app to control different motor attached to the [STM_Nucleo H743zi board](https://os.mbed.com/platforms/ST-Nucleo-H743ZI/) for which I've to set some GPIO Pins as output so that that I run them (by setting the GPIO pins as high) as per requirement.

I would like to configure multiple pins from the board for the motor input and output purpose but I'm having trouble configuring a single custom pin from the board which doesn't have any predefined aliases like the pwm_led0 in the zephyr samples.

I would like some help to find out the port number and pin number of the pins so that I can use [gpio_pin_configure()](https://docs.zephyrproject.org/latest/reference/peripherals/gpio.html#c.gpio_pin_configure) to set the corresponding pin (for e.g. PB_8 as an output pin)

For e.g. I would like to configure the PB_8 pin as per the pinout diagram attached in the figure below as an output pin.
![image](https://user-images.githubusercontent.com/32318187/120701604-a9721600-c4d0-11eb-9cf6-b51c8f6ee7e4.png)

The dts file provided by the vendor for the board is attached in [this folder](https://drive.google.com/drive/folders/1Vf9VEkpT_a9_drTIr7VKIisSCv4wlobR?usp=sharing) and the generated header file from the device tree build is [attached here](https://drive.google.com/file/d/1MAwBzoI9ulIzC08p91uBRYqn6yG9vr7j/view?usp=sharing).

This is the [code snippet](https://pastebin.com/itJnnUuS) I've used so far and I've created an [overlay file](https://pastebin.com/RbT8ZEKm)  for the same. There are some build errors I get if I use the files attached here. 

I'm new to zephyr so my approach may be wrong. Since there are not may tutorials available, I've written this code based on the examples from the samples folder.

Any help to resolve my question is appreciated.

Thanks in advance.

**Comment by galak** [comment]
@erwango can you take a look at this question.

**Comment by erwango** [comment]
@manuaatitya To help further, can you provide the build errors ?

**Comment by manuaatitya** [comment]
hi @erwango, 

thank you for the reply,

This is the error output I get, 

Configuring firmware for zephyr platform nucleo_h743zi
Using serial device.
Please check firmware/zephyr_apps/microros_extensions/microros_transports.h
for configuring serial device before build.
root@11aec4a083cc:/home/octobotics/uros_ws# ros2 run micro_ros_setup build_firmware.sh 
Crosscompiled environment: cleaning path
Building firmware for zephyr platform nucleo_h743zi
Selected app: motor_control
Configuration: Using transport-specific serial.conf
-- west build: generating a build system
-- Application: /home/octobotics/uros_ws/firmware/zephyr_apps/apps/motor_control
-- Zephyr version: 2.5.0 (/home/octobotics/uros_ws/firmware/zephyrproject/zephyr)
-- Found Python3: /usr/bin/python3.8 (found suitable exact version "3.8.5") found components: Interpreter 
-- Found west (found suitable version "0.11.0", minimum required is "0.7.1")
-- Board: nucleo_h743zi
-- Cache files will be written to: /root/.cache/zephyr
-- Using toolchain: zephyr 0.11.2 (/home/octobotics/uros_ws/firmware/zephyr-sdk)
-- Found dtc: /home/octobotics/uros_ws/firmware/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.5.0", minimum required is "1.4.6")
-- Found BOARD.dts: /home/octobotics/uros_ws/firmware/zephyrproject/zephyr/boards/arm/nucleo_h743zi/nucleo_h743zi.dts
-- Found devicetree overlay: /home/octobotics/uros_ws/firmware/zephyr_apps/apps/motor_control/boards/nucleo_h743zi.overlay
nucleo_h743zi.dts.pre.tmp:2746.17-2756.7: ERROR (phandle_references): /zephyr,user: Reference to non-existent node or label "gpio0"

nucleo_h743zi.dts.pre.tmp:2746.17-2756.7: ERROR (phandle_references): /zephyr,user: Reference to non-existent node or label "gpio0"

nucleo_h743zi.dts.pre.tmp:2746.17-2756.7: ERROR (phandle_references): /zephyr,user: Reference to non-existent node or label "gpio1"

nucleo_h743zi.dts.pre.tmp:2746.17-2756.7: ERROR (phandle_references): /zephyr,user: Reference to non-existent node or label "gpio0"

ERROR: Input tree has errors, aborting (use -f to force output)
CMake Error at /home/octobotics/uros_ws/firmware/zephyrproject/zephyr/cmake/dts.cmake:205 (message):
  command failed with return code: 2
Call Stack (most recent call first):
  /home/octobotics/uros_ws/firmware/zephyrproject/zephyr/cmake/app/boilerplate.cmake:535 (include)
  CMakeLists.txt:9 (include)


-- Configuring incomplete, errors occurred!
FATAL ERROR: command exited with status 1: /usr/bin/cmake -DWEST_PYTHON=/usr/bin/python3 -B/home/octobotics/uros_ws/firmware/build -S/home/octobotics/uros_ws/firmware/zephyr_apps/apps/motor_control -GNinja -DBOARD=nucleo_h743zi -DCONF_FILE=/home/octobotics/uros_ws/firmware/zephyr_apps/apps/motor_control/serial.conf '-GUnix Makefiles' -DCMAKE_VERBOSE_MAKEFILE=off -DMICRO_ROS_FIRMWARE_DIR=/home/octobotics/uros_ws/firmware -DMICRO_ROS_TRANSPORT=serial

**Comment by erwango** [comment]
For some reason, I'm not able to open nucleo_h743zi.dts file, but I'd assume it differs from upstream (https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/nucleo_h743zi/nucleo_h743zi.dts).

Though, reading the reported errors:
```
nucleo_h743zi.dts.pre.tmp:2746.17-2756.7: ERROR (phandle_references): /zephyr,user: Reference to non-existent node or label "gpio1"

nucleo_h743zi.dts.pre.tmp:2746.17-2756.7: ERROR (phandle_references): /zephyr,user: Reference to non-existent node or label "gpio0"
```
Please note that there is no `gpio0` or `gpio1` in STM32 .dtsi files.
So, I'd suggest to review these settings and have a new try with valid stm32 gpio nodes (&gpioa, &gpiob, ...)

**Comment by manuaatitya** [comment]
Hi @erwango, 
    I think, my question was not clearly understood, 

    
I need to access say pin 4, port a i.e. gpioa of the board (just as an example), and it has no special nodes like green led or buttons ... 

So if I have to use gpio_pin_configure() to set the pin 4, gpioa as output, how do I get the device struct since I don't have any device binding so I won't be to configure the pin without the device, i.e I wont be able to get the device like this

    const struct device *dev;		
    int ret;
    dev = device_get_binding(LED0);

So, basically my doubt is what should I do in place of 

    gpio_pin_configure( ?? , 14, GPIO_OUTPUT_HIGH);


Thanks in advance

If my method is wrong, please guide me as to how i can achieve the above.

**Comment by erwango** [comment]
>  it has no special nodes like green led or buttons

In that case, you need to add your own node for this pin, cf what is proposed below.
https://docs.zephyrproject.org/latest/guides/dts/bindings.html?highlight=inferred#inferred-bindings

Please note that you should use real references, such as (sticking to your example):
```
	zephyr,user {
		signal-gpios = <&gpioa 4 GPIO_ACTIVE_HIGH>;
	};
```

Then for configuring as OUTPUT, you can do:
```
   #include <drivers/gpio.h>

   const struct gpio_dt_spec signal =
           GPIO_DT_SPEC_GET(ZEPHYR_USER_NODE, signal_gpios);

   /* Configure the pin */
   gpio_pin_configure_dt(&signal, GPIO_OUTPUT_INACTIVE);

```

**Reply by manuaatitya** [reply]
Hi @erwango, 
     Thanks a lot for the reply, 
     When I execute the code that you referrred, I get the following error,
    
     error: variable 'signal' has initializer but incomplete type
     45 | const struct gpio_dt_spec signal = GPIO_DT_SPEC_GET(ZEPHYR_USER_NODE, signal_gpios);
     
I just have one more final doubt, 

In case, I have to add more custom pins, is it possible to add more gpio pins to this zephyr, user node like for e.g 

    zephyr,user {
		signal-gpios = <&gpioa 4 GPIO_ACTIVE_HIGH>;  // For port A pin 4
            signal-gpios = <&gpiob 0 GPIO_ACTIVE_HIGH>;   // For port B pin 0  ???
	};

I couldn't find more info on this in the documentation.

**Reply by tej113** [reply]
Hi @erwango can we multiple custom gpio inside zephyr,user device tree like below...

 zephyr,user {
    /* Define GPIO pins */
    my-gpio-pin-0 {
        signal-gpios = <&gpioa 4 GPIO_ACTIVE_HIGH>;
    };

    my-gpio-pin-1 {
        signal-gpios = <&gpioa 0 GPIO_ACTIVE_HIGH>;
    };

    my-gpio-pin-2 {
        signal-gpios = <&gpioa 1 GPIO_ACTIVE_HIGH>;
    };

    my-gpio-pin-3 {
        signal-gpios = <&gpioa 2 GPIO_ACTIVE_HIGH>;
    };
};


if this looks fine how to use them in main.c like binding pin configuration.

**Reply by tej113** [reply]
Hi @manuaatitya ,
Are you able to achieve declaring the multiple gpio pins

**Comment by erwango** [comment]
@manuaatitya 
You might need the include:
```
#include <drivers/gpio.h>
```

> In case, I have to add more custom pins,

It's possible, but you need to use a different name (than `signal-gpio`)

**Reply by manuaatitya** [reply]
Hi @erwango, 

     I've included the drivers/gpio header file in my code.
     
The complete error log is [attached here](https://pastebin.com/vcZFtiR8). 

The complete source code used is as [follows](https://pastebin.com/mMhX56A3)

I get error even after including the header file (which I had done previously also)

**Comment by erwango** [comment]
Have a try to move these lines inside the main functions:
```
   const struct gpio_dt_spec signal =
           GPIO_DT_SPEC_GET(ZEPHYR_USER_NODE, signal_gpios);

   /* Configure the pin */
   gpio_pin_configure_dt(&signal, GPIO_OUTPUT_INACTIVE);
```

**Reply by manuaatitya** [reply]
Hi @erwango,
I tried moving the instantiation and configuration codes within the main function, but I still get this [error log](https://pastebin.com/ai0GPEHb)

**Comment by erwango** [comment]
ok, you miss a:
`#define ZEPHYR_USER_NODE DT_PATH(zephyr_user)`

This was not documented, I'll take care to get it added.

**Reply by manuaatitya** [reply]
Hi @erwango,

   I still get this [error](https://pastebin.com/GXRgfcdK) after adding the # define line as you suggested.

**Reply by erwango** [reply]
It's working on my side.
So you probably miss the following:
```
 #include <zephyr.h>
 #include <device.h>
 #include <devicetree.h>
 #include <drivers/gpio.h>
```

**Comment by gabopushups** [comment]
This is some snippet from my dts, under the root node:

```
/ {
	model = "96Boards Gumstix AeroCore 2";
	compatible = "gumstix,aerocore2";

	chosen {
		zephyr,console = &usart1;
		zephyr,shell-uart = &usart1;
		zephyr,sram = &sram0;
		zephyr,flash = &flash0;
		zephyr,ccm = &ccm0;
	};
        mypins {
		compatible = "gpio-keys";
		dirA: button_A{
			gpios = <&gpioa 8 GPIO_ACTIVE_HIGH>;
			label = "PinDir 0";
		};
		dirB: button_B{
			gpios = <&gpioa 9 GPIO_ACTIVE_HIGH>;
			label = "PinDir 1";
	};
        aliases {
		ina = &dirA;
		inb = &dirB;
        };
}
```
As you can see, I created a node called "mypins", under which I used the binding "gpio-keys". In your dts. you'll need to change `gpioa 8` for `gpiob 8` and it should work for you, and use whatever aliases and names better suit your needs. 

Then in the app code I call the alias `ina` (for input A) under the name MPOS_NODE (for motor positive node, but as I said, use whatever name you need)

```
#define MPOS_NODE	DT_ALIAS(inb) // 
#if DT_NODE_HAS_STATUS(MPOS_NODE, okay)
#define MPOS_GPIO_LABEL	DT_GPIO_LABEL(MPOS_NODE, gpios)
#define MPOS_GPIO_PIN	DT_GPIO_PIN(MPOS_NODE, gpios)
#define MPOS_GPIO_FLAGS	(GPIO_INPUT | DT_GPIO_FLAGS(MPOS_NODE, gpios))
#else
#error "Unsupported board: INA devicetree alias is not defined"
#define MPOS_GPIO_LABEL	""
#define MPOS_GPIO_PIN	0
#define MPOS_GPIO_FLAGS	0
#endif
```
And finally

``` 
      const struct device *motPos;
       motPos = device_get_binding(MPOS_GPIO_LABEL);
	if (motPos == NULL) {
		printk("Error: didn't find %s device\n", MPOS_GPIO_LABEL);
		return;
	}
        int ret;
	ret = gpio_pin_configure(motPos, MPOS_GPIO_PIN, GPIO_OUTPUT_ACTIVE | MPOS_GPIO_FLAGS);
	if (ret != 0) {
		printk("Error %d: failed to configure %s pin %d\n",
		       ret, MPOS_GPIO_LABEL, MPOS_GPIO_PIN);
		return;
	}
```

**Reply by prajoshpremdas** [reply]
Hi @manuaatitya

I happened to stumble upon this post when the inferred-bindings seems not to work at my end.

My conclusion: 
GPIO_DT_SPEC_GET is added from the **zephyr-v2.6.0** tag only in the file _https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v2.6.0/include/drivers/gpio.h_. So please update your zephyr code base

Regards
Prajosh

**Reply by tej113** [reply]
> This is some snippet from my dts, under the root node:
> 
> ```
> / {
> 	model = "96Boards Gumstix AeroCore 2";
> 	compatible = "gumstix,aerocore2";
> 
> 	chosen {
> 		zephyr,console = &usart1;
> 		zephyr,shell-uart = &usart1;
> 		zephyr,sram = &sram0;
> 		zephyr,flash = &flash0;
> 		zephyr,ccm = &ccm0;
> 	};
>         mypins {
> 		compatible = "gpio-keys";
> 		dirA: button_A{
> 			gpios = <&gpioa 8 GPIO_ACTIVE_HIGH>;
> 			label = "PinDir 0";
> 		};
> 		dirB: button_B{
> 			gpios = <&gpioa 9 GPIO_ACTIVE_HIGH>;
> 			label = "PinDir 1";
> 	};
>         aliases {
> 		ina = &dirA;
> 		inb = &dirB;
>         };
> }
> ```
> 
> As you can see, I created a node called "mypins", under which I used the binding "gpio-keys". In your dts. you'll need to change `gpioa 8` for `gpiob 8` and it should work for you, and use whatever aliases and names better suit your needs.
> 
> Then in the app code I call the alias `ina` (for input A) under the name MPOS_NODE (for motor positive node, but as I said, use whatever name you need)
> 
> ```
> #define MPOS_NODE	DT_ALIAS(inb) // 
> #if DT_NODE_HAS_STATUS(MPOS_NODE, okay)
> #define MPOS_GPIO_LABEL	DT_GPIO_LABEL(MPOS_NODE, gpios)
> #define MPOS_GPIO_PIN	DT_GPIO_PIN(MPOS_NODE, gpios)
> #define MPOS_GPIO_FLAGS	(GPIO_INPUT | DT_GPIO_FLAGS(MPOS_NODE, gpios))
> #else
> #error "Unsupported board: INA devicetree alias is not defined"
> #define MPOS_GPIO_LABEL	""
> #define MPOS_GPIO_PIN	0
> #define MPOS_GPIO_FLAGS	0
> #endif
> ```
> 
> And finally
> 
> ```
>       const struct device *motPos;
>        motPos = device_get_binding(MPOS_GPIO_LABEL);
> 	if (motPos == NULL) {
> 		printk("Error: didn't find %s device\n", MPOS_GPIO_LABEL);
> 		return;
> 	}
>         int ret;
> 	ret = gpio_pin_configure(motPos, MPOS_GPIO_PIN, GPIO_OUTPUT_ACTIVE | MPOS_GPIO_FLAGS);
> 	if (ret != 0) {
> 		printk("Error %d: failed to configure %s pin %d\n",
> 		       ret, MPOS_GPIO_LABEL, MPOS_GPIO_PIN);
> 		return;
> 	}
> ```

This is for 1 gpio, 
similarly we need to define for how many gpios we are adding in devicetree file..???
If that is the case then code length will grow.

**Comment by slpp95prashanth** [comment]
Hi, I tried the gpio-nexus for adding J721E gpio in the device tree.

The device tree gets compiled. But the build fails while building blinky example.
Error:
error: device_dts_ord_DT_N_S_leds_S_led0_P_gpios_IDX_0_PH_ORD undeclared here (not in a function).
zephyr/build/zephyr/include/generated/devicetree_generated.h:441:29: error: DT_N_S_leds_S_led0_P_gpios_IDX_0_VAL_pin undeclared here (not in a function)

For neorv32 blinky example, The devicetree_generated.h has these definitions:
/* Generic property macros: */
#define DT_N_S_leds_S_led0_P_gpios_IDX_0_EXISTS 1
#define DT_N_S_leds_S_led0_P_gpios_IDX_0_PH DT_N_S_soc_S_gpio_S_gpio_ffffffc0
#define DT_N_S_leds_S_led0_P_gpios_IDX_0_VAL_pin 0
#define DT_N_S_leds_S_led0_P_gpios_IDX_0_VAL_pin_EXISTS 1
#define DT_N_S_leds_S_led0_P_gpios_IDX_0_VAL_flags 0
#define DT_N_S_leds_S_led0_P_gpios_IDX_0_VAL_flags_EXISTS 1
#define DT_N_S_leds_S_led0_P_gpios_FOREACH_PROP_ELEM(fn) fn(DT_N_S_leds_S_led0, gpios, 0)
#define DT_N_S_leds_S_led0_P_gpios_FOREACH_PROP_ELEM_SEP(fn, sep) fn(DT_N_S_leds_S_led0, gpios, 0)
#define DT_N_S_leds_S_led0_P_gpios_FOREACH_PROP_ELEM_VARGS(fn, ...) fn(DT_N_S_leds_S_led0, gpios, 0, VA_ARGS)
#define DT_N_S_leds_S_led0_P_gpios_FOREACH_PROP_ELEM_SEP_VARGS(fn, sep, ...) fn(DT_N_S_leds_S_led0, gpios, 0, VA_ARGS__)
#define DT_N_S_leds_S_led0_P_gpios_LEN 1
#define DT_N_S_leds_S_led0_P_gpios_EXISTS 1 

But for J721e (Which I am adding support) does not have these definitions in the devicetree_generated.h.

I have attached board dts (bbai_64.dts) and soc dts (tda4vm.dtsi). (edited)
[dts-files.zip](https://github.com/zephyrproject-rtos/zephyr/files/11676492/dts-files.zip)

**Reply by slpp95prashanth** [reply]
> Hi, I tried the gpio-nexus for adding J721E gpio in the device tree.
> 
> The device tree gets compiled. But the build fails while building blinky example. Error: error: device_dts_ord_DT_N_S_leds_S_led0_P_gpios_IDX_0_PH_ORD undeclared here (not in a function). zephyr/build/zephyr/include/generated/devicetree_generated.h:441:29: error: DT_N_S_leds_S_led0_P_gpios_IDX_0_VAL_pin undeclared here (not in a function)
> 
> For neorv32 blinky example, The devicetree_generated.h has these definitions: /* Generic property macros: */ #define DT_N_S_leds_S_led0_P_gpios_IDX_0_EXISTS 1 #define DT_N_S_leds_S_led0_P_gpios_IDX_0_PH DT_N_S_soc_S_gpio_S_gpio_ffffffc0 #define DT_N_S_leds_S_led0_P_gpios_IDX_0_VAL_pin 0 #define DT_N_S_leds_S_led0_P_gpios_IDX_0_VAL_pin_EXISTS 1 #define DT_N_S_leds_S_led0_P_gpios_IDX_0_VAL_flags 0 #define DT_N_S_leds_S_led0_P_gpios_IDX_0_VAL_flags_EXISTS 1 #define DT_N_S_leds_S_led0_P_gpios_FOREACH_PROP_ELEM(fn) fn(DT_N_S_leds_S_led0, gpios, 0) #define DT_N_S_leds_S_led0_P_gpios_FOREACH_PROP_ELEM_SEP(fn, sep) fn(DT_N_S_leds_S_led0, gpios, 0) #define DT_N_S_leds_S_led0_P_gpios_FOREACH_PROP_ELEM_VARGS(fn, ...) fn(DT_N_S_leds_S_led0, gpios, 0, VA_ARGS) #define DT_N_S_leds_S_led0_P_gpios_FOREACH_PROP_ELEM_SEP_VARGS(fn, sep, ...) fn(DT_N_S_leds_S_led0, gpios, 0, VA_ARGS__) #define DT_N_S_leds_S_led0_P_gpios_LEN 1 #define DT_N_S_leds_S_led0_P_gpios_EXISTS 1
> 
> But for J721e (Which I am adding support) does not have these definitions in the devicetree_generated.h.
> 
> I have attached board dts (bbai_64.dts) and soc dts (tda4vm.dtsi). (edited) [dts-files.zip](https://github.com/zephyrproject-rtos/zephyr/files/11676492/dts-files.zip)

FIxed: need to add gpio-nexus.yaml in the bindings.

**Comment by Madhu1524** [comment]


**Reply by erwango** [reply]
Please open a dedicated issue, this has nothing to do with current topic which was closed several years ago on a different platform
